apps/backend/app/services/model_training/feature_engineer.py

```
@@ -16,12 +18,101 @@
+            serialized expression tree (including legacy raw Python code)
+    """
+    try:
+        data = json.loads(definition_code)
+    except (json.JSONDecodeError, TypeError):
+        raise UnsafeFeatureDefinitionError(
+            message=(
+                "Feature definition must be a JSON-serialized expression tree. "
+                "Raw code is not supported — recreate the feature with the "
+                "Visual Feature Builder."
+            )
+        )
```

Ruff (0.15.15)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)